// @tpio/javascript-opensdk/dist/src/rest/messages/sessionRequest.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const configHelper_1 = __importDefault(require("../../sdk/internal/helpers/configHelper"));
class SessionRequest {
    constructor(reportSettings, capabilities) {
        this.reportSettings = reportSettings;
        this.capabilities = capabilities;
        this.language = 'Nodejs';
        this.sdkVersion = configHelper_1.default.sdkVersion();
    }
    capsAsObject() {
        const caps = this.capabilities;
        const m = {};
        caps.map_.forEach((value, key) => {
            if (!key.startsWith('_tp')) {
                m[key] = value;
            }
        });
        return m;
    }
    toObject() {
        return {
            language: this.language,
            sdkVersion: this.sdkVersion,
            capabilities: this.capsAsObject(),
            projectName: this.reportSettings.projectName,
            jobName: this.reportSettings.jobName,
            reportType: this.reportSettings.reportType,
            reportName: this.reportSettings.reportName,
            reportPath: this.reportSettings.reportPath,
        };
    }
}
exports.default = SessionRequest;
